### ISSUE TITLE
How to evaluate loss in.val mode

### ISSUE THREAD

**Issue by Funcery** [ultralytics/ultralytics#13601]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Due to a special task, I need to evaluate the loss in val mode, I tried to query all the parameters and return values in val, and none of them got a result. I also couldn't accomplish this by setting the learning rate to zero in training mode.
really need help
thanks

**Comment by glenn-jocher** [comment]
@Funcery hello,

Thank you for reaching out with your question! To evaluate the loss during the validation phase, you can modify the validation loop to include loss computation. Unfortunately, the default `val` mode in YOLOv8 does not return the loss directly, but you can achieve this by customizing the validation script.

Here's a concise example of how you can modify the validation process to include loss evaluation:

```python
from ultralytics import YOLO

# Load your model
model = YOLO("path/to/your/model.pt")

# Define your custom validation function
def custom_val(model, data, device='cuda'):
    # Load the validation dataset
    dataloader = model.val_dataloader(data=data, batch_size=16, device=device)
    
    # Set the model to evaluation mode
    model.model.eval()
    
    # Initialize loss
    total_loss = 0.0
    num_batches = 0
    
    # Iterate over the validation dataset
    for batch in dataloader:
        images, targets = batch['img'].to(device), batch['labels'].to(device)
        
        # Forward pass
        with torch.no_grad():
            outputs = model.model(images)
        
        # Compute loss
        loss, _ = model.loss(outputs, targets)
        total_loss += loss.item()
        num_batches += 1
    
    # Calculate average loss
    avg_loss = total_loss / num_batches
    print(f'Validation Loss: {avg_loss:.4f}')
    return avg_loss

# Run custom validation
custom_val(model, data='path/to/your/validation/data.yaml')
```

This script sets up a custom validation loop that computes the loss for each batch and averages it over the entire validation dataset. Make sure to adjust the paths to your model and data accordingly.

If you encounter any issues or need further assistance, please ensure you are using the latest versions of `torch` and `ultralytics` packages. You can update them using the following commands:

```bash
pip install --upgrade torch ultralytics
```

For more detailed information on model validation, you can refer to our [documentation](https://docs.ultralytics.com/modes/val/).

I hope this helps! If you have any further questions, feel free to ask.
